### ISSUE TITLE
BUG: scipy.linalg.blas.dnrm2 may return error result when incx < 0

### ISSUE THREAD

**Issue by M0reDr1nk** [scipy/scipy#16930]
### Describe your issue.

scipy.linalg.blas.dnrm2 may returns error result when incx<0 :
input array:[10, 10, 10, 10, 10, 10, 10, 10, 10]
n:5
offx:3
incx:-1
the scipy.linalg.blas.dnrm2 function returns result 20.0

If we calculate the result follow the document in netlib:  
https://netlib.org/lapack/explore-html/df/d28/group__single__blas__level1_gab5393665c8f0e7d5de9bd1dd2ff0d9d0.html

we will get result sqrt(500) -> 22.36

### Reproducing Code Example

```python
import os
import time
import numpy as np
from scipy import linalg

x = [10,10,10,10,10,10,10,10,10]
n = 5
offx = 3
incx = -1
print(f"x:{x}")
print(f"n:{n}")
print(f"offx:{offx}")
print(f"incx:{incx}") 
res = linalg.blas.dnrm2(x, n, offx, incx)
print(f"res:{res}")
```

### Error message

```shell
x:[10, 10, 10, 10, 10, 10, 10, 10, 10]
n:5
offx:3
incx:-1
res:20.0
```

### SciPy/NumPy/Python version information

1.7.3 1.21.5 sys.version_info(major=3, minor=7, micro=5, releaselevel='final', serial=0)

**Comment by M0reDr1nk** [comment]
run the same code on Kaggle Notebooks, i get the result here:
1.7.3 1.21.6 sys.version_info(major=3, minor=7, micro=12, releaselevel='final', serial=0)
x:[10, 10, 10, 10, 10, 10, 10, 10, 10]
n:5
offx:3
incx:-1
res:0.0

**Comment by steppi** [comment]
@M0reDr1nk this was fixed in OpenBLAS by https://github.com/xianyi/OpenBLAS/pull/4190. I’ll leave this issue open until the OpenBLAS wheel included with SciPy has the fix.

**Comment by ev-br** [comment]
> leave this issue open until the OpenBLAS wheel included with SciPy has the fix.

The linked OpenBLAS PR has the milestone 0.3.24. I wonder what is the lowest version for SciPy wheels (and conda-forge binaries maybe?). @tylerjereddy could you advise?

**Comment by lucascolley** [comment]
> what is the lowest version for SciPy wheel

looks like `0.3.21.dev` if I'm looking in the right place: https://github.com/scipy/scipy/blob/2aee5efcbe3720f41fe55f336f492ae0acbecdee/tools/openblas_support.py#L16-L17

**Comment by tylerjereddy** [comment]
working on it, given that the "new way" in https://github.com/scipy/scipy/pull/20074 may need a bit more time

**Comment by tylerjereddy** [comment]
This problem actually seems quite nasty to solve as outlined in the cross-linked PR. Basically, the old OpenBLAS (before scipy-openblas32 infrastructure transition project) binaries are too old to fix the failing test case. The new ones present the usual symbol/linking issues that, if solved, would mean Matti's stuff could move forward anyway I think.

**Comment by rgommers** [comment]
Thanks for trying Tyler. Looks like we'll just have to work through the symbol renaming stuff - it's not that far off.

**Comment by tylerjereddy** [comment]
The bump to OpenBLAS `0.3.26`, with Matti's help, in gh-20215, seems to have fixed the issue only on a subset of architectures/platforms. So, what will probably happen is leaving this issue open and xfailing the new test for now.

**Comment by mattip** [comment]
Maybe worth re-raising with OpenBLAS. Since this test is a direct call to dnrm2, would it be "easy" to debug there?

**Comment by tylerjereddy** [comment]
I opened the issue upstream (see above) and offered to help Martin with the regression test.

**Comment by tylerjereddy** [comment]
Sounds good, thanks for adding upstream regression test.

**Comment by mattip** [comment]
FWIW, I built tarballs from the latest OpenBLAS HEAD including the fix for this on arm64/aarch64. I think this diff to `tools/openblas_support.py` should work
```diff
diff --git a/tools/openblas_support.py b/tools/openblas_support.py
index 241d5be51..9de33bec1 100644
--- a/tools/openblas_support.py
+++ b/tools/openblas_support.py
@@ -13,8 +13,8 @@ from tempfile import mkstemp, gettempdir
 from urllib.request import urlopen, Request
 from urllib.error import HTTPError
 
-OPENBLAS_V = '0.3.26'
-OPENBLAS_LONG = 'v0.3.26'
+OPENBLAS_V = '0.3.26.dev'
+OPENBLAS_LONG = 'v0.3.26-322-g4a6025af'
 BASE_LOC = 'https://anaconda.org/multibuild-wheels-staging/openblas-libs'
 NIGHTLY_BASE_LOC = (
     'https://anaconda.org/scientific-python-nightly-wheels/openblas-libs'

```

**Comment by rgommers** [comment]
No one is proposing to bump the minimum supported LAPACK version I think? We could bump it to 3.8 or 3.9 if there are new APIs from those releases we want to include I think.

**Comment by ilayn** [comment]
Sure but negative incx functionality is from 3.10. I don't have any objections to it to be honest it doesn't seem like folks are using system defaults and install openblas anyways

**Comment by ilayn** [comment]
Let me try again; LAPACK/BLAS API never changes. That's written in stone. But if someone relies on behavior of `incx<0` or pins OpenBLAS version you are effectively way past 3.7.1 already. If things break downstream because of this, it is a **very** nasty bug to figure out because our `linalg.norm` depends on this.

- So whatever Fujitsu or other folks are experimenting with LAPACK is overridden by our OpenBLAS requirement. And since they are not coming with pitchforks our way, I think they are keeping things up-to-date too. Also if it is new, then we are safe already but we definitely force newer LAPACK through OpenBLAS versions. Following back from a few versions is not a problem but if we are dropping support for older Python versions there is not much evidence in favor of holding support for LAPACK versions too older than that.

We bumped to 3.7 also by guesstimate and not a rigorous analysis either so if we are doing it let's do it properly. LAPACK version is a fundamental upgrade for any number crunching system anyways so folks also have a stake at keeping it up-to-date.

**Comment by rgommers** [comment]
Okay then raise a clear error when `incx` is negative.

We cannot aggressively bump LAPACK the moment one library supports something new without a very compelling reason - even if that is our default library. The fix is in OpenBLAS 0.3.24, which we support in our own wheel builds since this week. The fix is not in for example Ubuntu 22.04 LTS ([that's at 0.3.20](https://packages.ubuntu.com/jammy/libopenblas-dev)), which is the default CI image for GitHub Actions ([ubuntu-latest](https://github.com/actions/runner-images/?tab=readme-ov-file#available-images)).

Even if we were to do this, there is a very practical issue: we don't check the supported LAPACK version we're building against. We rely on getting build errors - which makes perfect sense when wrapping a new API. It doesn't work though if it's an existing API that _changes behavior_. In this case, `incx<0` changed from returning incorrect values to correct ones. Other libraries may error for negative input, or crash, or whatever - we don't know at build time.

**Comment by ilayn** [comment]
True, my emphasis is on the fact that while we are defensive about LAPACK version, which perfectly fine, we should be careful about indirect consequences of bumping other things up. 

Otherwise I don't think we have urgency or good reasons to bump up all the way (though 3.10 included Anderson's safety scaling paradigm which are really good but they can wait).
